File: 2/intcodes.cpp
#include <iostream>
#include <fstream>
#include <vector>

int totalise_instructions(std::vector<int> codes, int noun, int verb) {
    codes[1] = noun;
    codes[2] = verb;

    for (auto i = 0; i < codes.size(); i += 4) {
        auto op = codes[i];
        auto first = codes[i+1];
        auto second = codes[i+2];
        auto pos = codes[i+3];
        if (op == 99) {
            return codes[0];
        }
        codes[pos] = (op&1) ? codes[first]+codes[second] : codes[first]*codes[second];
    }
}

int main() {
    std::ifstream f{"input.txt"};
    std::string line;
    std::vector<int> intcodes{};

    while (std::getline(f, line, ',')) {
            intcodes.push_back(std::stoi(line));
    }
    std::cout << "Part 1: " << totalise_instructions(intcodes, 12, 2) << '\n';

    for (auto a = 0; a < 100; a++) {
        for (auto b = 0; b < 100; b++) {
            if (totalise_instructions(intcodes, a, b) == 19690720) {
                std::cout << "Part 2: " << 100*a + b << std::endl;
                return 0;
            }
        }
    }
}

File: 2/intcodes.py

import itertools
intcodes = [int(x) for x in open("input.txt").readline().rstrip("\n").split(",")]

def totalise_instructions(codes, cell_one, cell_two):
    codes[1:3] = cell_one, cell_two
    i = 0
    while i < len(codes):
        op, first, second, pos = codes[i:i+4]
        if op == 99:
            return codes[0]
        codes[pos] = codes[first]+codes[second] if op%2 else codes[first]*codes[second]
        i += 4

print("Part 1:", totalise_instructions(intcodes[:], 12, 2))

for a, b in itertools.permutations(range(100), 2):
        if totalise_instructions(intcodes[:], a, b) == 19690720:
            print("Part 2:", 100*a + b)
            break
